### PR TITLE
Add support for variadic dispatch

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds support for variadic dispatch using syntax from #3

### CODE REVIEW

**Comment by cpcloud** [comment]
FYI, still a WIP. There's currently an issue ensuring that variadic signature "lengths" are comparable to partially, or non-variadic functions.

For example,

```python
@dispatch(str, int)
def foo(a, b):
    return int(a) + b

@dispatch([object])
def foo(*args):
    return sum(map(int, args))
```

The second `foo` should supercede the first, but because `supercede` checks that the length of two signatures are the same, it never does.

**Comment by cpcloud** [comment]
@mrocklin @shoyer @llllllllll @mariusvniekerk

Any chance I could get one of you to review this?

cc @Andrew-peng, who addressed the issue from my previous comment

File: multipledispatch/tests/test_conflict.py [hunk 1]
```diff
@@ -60,3 +61,46 @@ def test_ordering():
...
     assert super_signature([[object], [type]]) == [type]
+
+
+def test_supercedes_variadic():
+    assert supercedes((Variadic[B],), (Variadic[A],))
+    assert supercedes((B, Variadic[A]), (Variadic[A],))
+    assert supercedes((Variadic[A],), (Variadic[(A, C)],))
+    assert supercedes((A, B, Variadic[C]), (Variadic[object],))
+    assert supercedes((A, Variadic[B]), (Variadic[A],))
+    assert supercedes(tuple([]), (Variadic[A],))
```
Comment (llllllllll) [inline multipledispatch/tests/test_conflict.py]: I actually really appreciate not using a tuple literal here. It makes this easier to read, thanks.

File: multipledispatch/tests/test_conflict.py [hunk 2]
```diff
@@ -60,3 +61,46 @@ def test_ordering():
...
+    assert supercedes((A, A, A), (A, Variadic[A]))
+    assert not supercedes((Variadic[A],), (Variadic[B],))
```
Comment (llllllllll) [inline multipledispatch/tests/test_conflict.py]: could you add a newline to split these or break it into two tests?

File: multipledispatch/conflict.py [hunk 3]
```diff
@@ -1,19 +1,73 @@
 from .utils import _toposort, groupby
+from .variadic import isvariadic
+
 
 class AmbiguityWarning(Warning):
     pass
 
 
 def supercedes(a, b):
```
Comment (llllllllll) [inline multipledispatch/conflict.py]: Should we fix this typo? It's not new to this PR but I noticed it while reviewing the tests.

File: multipledispatch/tests/test_conflict.py (hunk 2 above)
Comment (llllllllll) [inline multipledispatch/tests/test_conflict.py]: actually, it looks like these are just flipped arguments from the positive tests, should we write a little helper like:

```python
def assert_supersedes(a, b):
    assert supercedes(a, b)
    assert not supercedes(b, a)
```

and then write each pair once

File: multipledispatch/tests/test_dispatcher.py [hunk 4]
```diff
@@ -258,3 +261,139 @@ def _(a):
...
+
+    @f.register([float])
+    def _1(*args):
+        return args
+
+    assert f(1.0) == (1.0,)
+    assert f(1.0, 2.0, 3.0) == (1.0, 2.0, 3.0)
+
+
+def test_vararg_dispatch_ambiguity():
```
Comment (llllllllll) [inline multipledispatch/tests/test_dispatcher.py]: we might want to add a tests where the ambiguity is in the variadic itself like:
```
In [3]: @f.register(object, [int])
   ...: def _(*args):
   ...:     pass
   ...: 

In [4]: @f.register(int, [object])
   ...: def _(*args):
   ...:     pass
   ...: 

```

File: multipledispatch/tests/test_dispatcher.py [hunk 5]
```diff
@@ -258,3 +261,139 @@ def _(a):
...
+    f = Dispatcher('f')
+    assert raises(TypeError, lambda: f.register([float], str)(lambda: None))
+
+
+def test_vararg_has_multiple_elements():
+    f = Dispatcher('f')
+    assert raises(TypeError, lambda: f.register([float, str])(lambda: None))
+
+
+def test_vararg_dispatch_simple():
```
Comment (llllllllll) [inline multipledispatch/tests/test_dispatcher.py]: should we add a test where we have a vararg registered and pass incompatible arguments to make sure it still gives us a notimplemented?

File: bench/test_simple.py [hunk 6]
```diff
@@ -1,14 +1,22 @@
 from multipledispatch import dispatch
 
+try:
+    range = xrange
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply bench/test_simple.py]: Indeed. I added six after adding this :)

File: multipledispatch/conflict.py (hunk 3 above)
Comment (cpcloud) [reply multipledispatch/conflict.py]: Looks like it's a legitimate spelling: https://english.stackexchange.com/questions/305034/supercede-or-supersede

I guess we have some linguistic contrarians working on multiple dispatch :)

File: multipledispatch/dispatcher.py [hunk 7]
```diff
@@ -46,6 +46,64 @@ def restart_ordering(on_ambiguity=ambiguity_warn):
...
+            Get the next signature
+    """
+    sigiter = iter(full_signature)
+    sig = next(sigiter)
+    for typ in types:
+        matches = issubclass(typ, sig)
+        if isvariadic(sig):
+            # if the current type matches the current variadic signature yield
+            # True
+            if matches:
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply multipledispatch/dispatcher.py]: Yep, thanks for pointing that out.

File: multipledispatch/conflict.py (hunk 3 above)
Comment (llllllllll) [reply multipledispatch/conflict.py]: nice citation, my mistake.

File: multipledispatch/dispatcher.py [hunk 8]
```diff
@@ -164,16 +222,35 @@ def add(self, signature, func):
...
 
-        self.funcs[signature] = func
+            # handle variadic signatures
+            if isinstance(typ, list):
+                if index != len(signature):
+                    raise TypeError(
+                        'Variadic signature must be the last element'
+                    )
+
+                if len(typ) != 1:
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply multipledispatch/dispatcher.py]: This was to maintain consistency with the current spelling of union types in the rest of the library.

File: multipledispatch/dispatcher.py [hunk 9]
```diff
@@ -260,11 +337,16 @@ def dispatch(self, *types):
             return None
 
     def dispatch_iter(self, *types):
+
         n = len(types)
         for signature in self.ordering:
             if len(signature) == n and all(map(issubclass, types, signature)):
                 result = self.funcs[signature]
                 yield result
+            elif any(isvariadic(sig) for sig in signature):
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply multipledispatch/dispatcher.py]: Yep, makes sense.

File: multipledispatch/tests/test_conflict.py (hunk 2 above)
Comment (cpcloud) [reply multipledispatch/tests/test_conflict.py]: Sounds good.

File: multipledispatch/tests/test_dispatcher.py (hunk 5 above)
Comment (cpcloud) [reply multipledispatch/tests/test_dispatcher.py]: For sure.

File: multipledispatch/tests/test_dispatcher.py (hunk 4 above)
Comment (cpcloud) [reply multipledispatch/tests/test_dispatcher.py]: Agreed.

File: multipledispatch/variadic.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
+import six
+
+from .utils import typename
+
+
+class VariadicSignatureType(type):
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply multipledispatch/variadic.py]: Nice catch, thanks.

File: multipledispatch/utils.py [hunk 11]
```diff
@@ -1,3 +1,6 @@
+from collections import OrderedDict
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply multipledispatch/utils.py]: This was to make testing reliable. We can revert back to `dict` but it's then possible to introduce bugs in cases like

```python
@f.register(str, [object])
@f.register(object, [object])
```

if the signature compatibility check is ever changed.

File: multipledispatch/utils.py [hunk 12]
```diff
@@ -45,18 +48,19 @@ def _toposort(edges):
     [2] http://en.wikipedia.org/wiki/Toposort#Algorithms
     """
     incoming_edges = reverse_dict(edges)
-    incoming_edges = dict((k, set(val)) for k, val in incoming_edges.items())
-    S = set((v for v in edges if v not in incoming_edges))
+    incoming_edges = OrderedDict((k, set(val))
+                                 for k, val in incoming_edges.items())
+    S = OrderedDict.fromkeys((v for v in edges if v not in incoming_edges))
```
Comment (cpcloud) [inline multipledispatch/utils.py]: @Andrew-peng we can remove one level of parentheses here.

File: multipledispatch/variadic.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    return isinstance(obj, VariadicSignatureType)
+
+
+class VariadicSignatureMeta(type):
+    """A metaclass that overrides ``__getitem__`` on the class. This is used to
+    generate a new type for Variadic signatures. See the Variadic class for
+    examples of how this behaves.
+    """
+    def __getitem__(self, variadic_type):
+        assert isinstance(variadic_type, (type, tuple)), type(variadic_type)
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply multipledispatch/variadic.py]: Indeed, we'll make that change.

File: multipledispatch/utils.py (hunk 12 above)
Comment (Andrew-peng) [reply multipledispatch/utils.py]: 👍

File: multipledispatch/variadic.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Andrew-peng) [reply multipledispatch/variadic.py]: 👍

File: multipledispatch/tests/test_dispatcher.py (hunk 4 above)
Comment (Andrew-peng) [reply multipledispatch/tests/test_dispatcher.py]: Added

File: multipledispatch/tests/test_dispatcher.py (hunk 5 above)
Comment (Andrew-peng) [reply multipledispatch/tests/test_dispatcher.py]: Added

File: multipledispatch/dispatcher.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Andrew-peng) [reply multipledispatch/dispatcher.py]: Could also do 
```python
yield matches
if not isvariadic(sig):
    sig = next(sigiter)
```

File: bench/test_simple.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply bench/test_simple.py]: @Andrew-peng can you change this to `range = six.moves.xrange`?

File: multipledispatch/dispatcher.py [hunk 14]
```diff
@@ -46,6 +46,59 @@ def restart_ordering(on_ambiguity=ambiguity_warn):
...
+        matches = issubclass(typ, sig)
+        if isvariadic(sig):
+            # if the current type matches the current variadic signature yield
+            # True, else yield False
+            yield matches
+        else:
+            # we're not matching a variadic argument, so move to the next
+            # element in the signature
+            yield matches
+            sig = next(sigiter)
```
Comment (cpcloud) [inline multipledispatch/dispatcher.py]: @Andrew-peng I like your suggestion of

```python
yield matches
if not isvariadic(sig)
    sig = next(sigiter)
```

File: multipledispatch/dispatcher.py [hunk 15]
```diff
@@ -260,11 +332,16 @@ def dispatch(self, *types):
             return None
 
     def dispatch_iter(self, *types):
+
         n = len(types)
         for signature in self.ordering:
             if len(signature) == n and all(map(issubclass, types, signature)):
                 result = self.funcs[signature]
                 yield result
+            elif len(signature) and isvariadic(signature[-1]):
```
Comment (cpcloud) [inline multipledispatch/dispatcher.py]: I believe this is guaranteed to be a `tuple`, in which case you can reduce to `elif signature and isvariadic(signature[-1]):`

You might even be able to reduce to just `isvariadic(signature[-1])` since I don't multiple dispatch allows an empty signature which means `signature[-1]` is always valid.

File: multipledispatch/tests/test_variadic.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class A(object): pass
+class B(A): pass
+class C(object): pass
+
+
+def test_is_variadic():
+	assert isvariadic(Variadic[A])
+	assert not isvariadic(A)
```
Comment (cpcloud) [inline multipledispatch/tests/test_variadic.py]: Indentation is looking a little weird here.

File: multipledispatch/dispatcher.py (hunk 15 above)
Comment (llllllllll) [reply multipledispatch/dispatcher.py]: I believe you can have an empty signature. This PR also adds tests that assert that explicitly empty supersedes a variadic.

File: multipledispatch/variadic.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        Returns
+        -------
+        bool
+            Whether or not `other` is equal to `self`
+        """
+        return (isvariadic(other) and
+                set(self.variadic_type) == set(other.variadic_type))
+
+    def __hash__(self):
+        return hash((type(self), frozenset(self.variadic_type)))
```
Comment (cpcloud) [inline multipledispatch/variadic.py]: No need to call `frozenset` here since `self.variadic_type` is already a `tuple` which is hashable.

Comment (llllllllll) [reply multipledispatch/variadic.py]: do we sort this tuple in any way? I believe we want `Variadic[int, float]` and `Variadic[float, int]` to resolve to the same thing.

File: multipledispatch/dispatcher.py (hunk 15 above)
Comment (cpcloud) [reply multipledispatch/dispatcher.py]: Interesting, it's definitely possible to define a function with an empty signature.

File: multipledispatch/variadic.py (hunk 17 above)
Comment (cpcloud) [reply multipledispatch/variadic.py]: Ah, great point. This is fine then.

File: multipledispatch/dispatcher.py (hunk 14 above)
Comment (Andrew-peng) [reply multipledispatch/dispatcher.py]: 👍 edited

File: multipledispatch/tests/test_variadic.py (hunk 16 above)
Comment (Andrew-peng) [reply multipledispatch/tests/test_variadic.py]: defaulted to tabs on accident

File: multipledispatch/variadic.py (hunk 17 above)
Comment (Andrew-peng) [reply multipledispatch/variadic.py]: 👍  I used `frozenset` so that `Variadic[(int, float)]` would match `Variadic[(float, int)]`

**Comment by cpcloud** [comment]
@Andrew-peng thanks for adding the benchmark

File: bench/test_simple.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply bench/test_simple.py]: @Andrew-peng not sure if you saw this comment

[the remark this answers is not in the document]
Comment (cpcloud) [reply bench/test_simple.py]: I'll fix this up, don't worry about it

**Comment by cpcloud** [comment]
@llllllllll @mrocklin Any more thoughts here?

File: multipledispatch/dispatcher.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (llllllllll) [reply multipledispatch/dispatcher.py]: cool, we can always decide to relax this later.

**Comment by llllllllll** [comment]
+1 from me also, I want to give people a chance to comment on the introduction of six, but otherwise it seems good to merge.

Thanks!

**Comment by cpcloud** [comment]
@llllllllll Awesome thanks!

I'm also happy to inline the `with_metaclass` code directly from six if that's what folks would prefer.

**Comment by cpcloud** [comment]
Updated the docs and the readme

**Comment by mrocklin** [comment]
> +1 from me also, I want to give people a chance to comment on the introduction of six, but otherwise it seems good to merge.

If folks are waiting for me to chime in here I recommend not waiting.  I'm happy to defer to @llllllllll and @cpcloud 's judgment

**Comment by llllllllll** [comment]
Okay, merging! Thanks @Andrew-peng and @cpcloud.
